fmn/sse/server.py

```diff
@@ -0,0 +1,301 @@
+        cc = protocol.ClientCreator(
+            reactor,
+            twisted_connection.TwistedProtocolConnection,
+            self.amqp_parameters,
+        )
+        self._deferred_connection = cc.connectTCP(self.amqp_host, self.amqp_port)
+        self._deferred_connection.addCallback(lambda conn: conn.ready)
+        self.connection = None
+        self.channel = None
+
+        # Maps queues to open requests. The datastructure is in the format
+        # {
+        #   RequestKey: {'queue': RabbitQueue, 'requests': []},
+        # }
+        self.subscribers = {}
+
+    @defer.inlineCallbacks
+    def queue(self, exchange=None, queue=None, routing_key=None,
+              queue_arguments=None):
+        """
+        Construct and retrieve a queue from RabbitMQ.
+
+        This will declare the exchange and queue in RabbitMQ if they don't
+        already exist.
+
+        The queue returned is a subclass of twisted.internet.defer.DeferredQueue.
+        Calling ``get`` on the queue returns a ``Deferred`` that fires with
+        (channel, method, properties, body) where the types are
+        (pika.Channel, pika.spec.Basic.Deliver, pika.spec.BasicProperties, bytes).
+
+        Afterwards you can do channel.basic_ack(delivery_tag=method.delivery_tag)
+        to acknowledge message reciept.
+
+        :param exchange:        The exchange to declare and bind the queue to.
+        :type  exchange:        unicode str
+        :param queue:           The queue to declare, bind to the exchange,
+                                and consume from.
+        :type  queue:           unicode str
+        :param routing_key:     The queue routing key.
+        :type  routing_key:     unicode str
+        :param queue_arguments: custom arguments for the queue's construction
+        :type  queue_agrements: dict
+
+        :return: A Deferred that fires with a tuple of the queue to consume
+                 messages from and a consumer tag.
+        :rtype: Deferred of (pika.ClosableDeferredQueue, unicode str)
+        """
+        # TODO Handle errors appropriately
+        if not self.connection:
+            self.connection = yield self._deferred_connection
+        if not self.channel:
+            self.channel = yield self.connection.channel()
+
+        # TODO Consider using ``passive=True`` on ``exchange_declare`` and
+        # ``queue_declare``. This causes an error if the queue or exchange don't
+        # already exist, which we could then turn into a HTTP 404. Allowing
+        # requests to cause exchanges or queues to be created seems like an easy
+        # DoS attack.
+        yield self.channel.exchange_declare(exchange=exchange)
+        yield self.channel.queue_declare(
+            queue=queue,
+            durable=True,
+            auto_delete=False,
+            arguments={'x-message-ttl': self.expire_ms}
+        )
+        # TODO investigate why this routing_key is used
```

Comment:
That makes sense, but does this work? The queue name is still the same, so currently what happens is if you request ``/user/jcline/`` and ``/group/jcline/`` you get a queue, ``jcline`` with two routing keys: ``user-jcline`` and ``group-jcline``.

I'm still pretty new working with AMQP, but doesn't that mean any message with either of those routing keys will be routed to the queue? I'd get both the group messages and user messages in one queue, right?